### PR TITLE
fix(sidebaritem): fix sidebar item with next link

### LINKED ISSUE #430: Maximum update depth exceeded with Next/Link and SidebarItem
**Describe the bug**
I opened a bug ticket here, see: https://github.com/vercel/next.js/issues/42902

**To Reproduce**
See: https://codesandbox.io/p/sandbox/hopeful-thompson-1lw449

**Expected behavior**
That it only renders once.

**Additional context**
This problem is occuring, because we're currently creating Components inside of other Components for e. g.: https://github.com/themesberg/flowbite-react/blob/v0.3.4/src/lib/components/Sidebar/SidebarItem.tsx#L34-L56

### PR DESCRIPTION
**Pull request by myasteiner**
## Description

This PR fixes the `next/link` component for the `SidebarItem`.

Fixes #430 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by rluders** [comment]
I'm not sure about this one... maybe @tulup-conner could give it a look.

**Comment by tulup-conner** [comment]
I don't totally understand what this does TBH. I don't really use Next. I don't think it makes sense for the library to have Next-specific patches.

**Comment by rluders** [comment]
Yes. I agree with you one this one @tulup-conner, I don't want to see specific-nextjs or any specific stuff landing at the flowbite-react. I have the same feeling that this is too specific. I'm wondering if #430 can't be solved in some more generic approach.

**Comment by rluders** [comment]
Also, I'm seeing the "Legacy" word there... So, is this for new Next.js version or just to support oldest versions? I mean, if this is a legacy support, I'll say: we don't care.

File: src/lib/components/Footer/FooterLink.tsx [hunk 1]
```diff
@@ -1,19 +1,21 @@
...
 
-export const FooterLink: FC<FooterLinkProps> = ({ children, className, href }) => {
+export const FooterLink = forwardRef<HTMLAnchorElement, FooterLinkProps>(({ children, className, href }, ref) => {
   const theme = useTheme().theme.footer.groupLink.link;
 
   return (
     <li className={classNames(theme.base, className)}>
-      <a href={href} className={theme.href}>
+      <a href={href} className={theme.href} ref={ref}>
         {children}
```
Comment (rluders) [inline src/lib/components/Footer/FooterLink.tsx:15]: BTW, I'm fine to have the `ref=` here. But I'm not sure about the remaining Next.js specific stuff.

File: src/lib/components/Navbar/NavbarLink.tsx [hunk 2]
```diff
@@ -8,26 +8,31 @@ export interface NavbarLinkProps extends PropsWithChildren<ComponentProps<'a'>>
...
+          className={classNames(
+            theme.base,
+            {
+              [theme.active.on]: active,
+              [theme.active.off]: !active && !disabled,
+            },
+            theme.disabled[disabled ? 'on' : 'off'],
+            className,
+          )}
+          ref={ref}
```
Comment (rluders) [inline src/lib/components/Navbar/NavbarLink.tsx:28]: Same stuff here.

File: src/lib/components/Next/NextLinkLegacyBehavior.tsx [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import { useCallback, useState, type FC, type ReactElement } from 'react';
```
Comment (rluders) [inline src/lib/components/Next/NextLinkLegacyBehavior.tsx:1]: This component seems to be like a "documented solution" not a flowbite-react component. So, maybe we could add it to the documentation insted? Some "Known Issues" section?

File: src/lib/components/Sidebar/SidebarItem.tsx [hunk 4]
```diff
@@ -31,32 +88,8 @@ const SidebarItem = forwardRef<Element, SidebarItemProps>(
...
-        data-testid="flowbite-sidebar-item-content"
-        id={`flowbite-sidebar-item-${id}`}
-      >
-        {children}
-      </span>
-    );
-
     return (
-      <ListItem>
+      <ListItem id={id} isCollapsed={isCollapsed} contentChildren={children}>
```
Comment (rluders) [inline src/lib/components/Sidebar/SidebarItem.tsx:92]: I felt like having the same `id` there and at line 144 will be a problem. I mean, the `id` should be always unique. Right? Maybe adding some suffix to the one at 144 line, something like `${id}-children`.

**Review by rluders — CHANGES_REQUESTED** [review]

**Comment by esko22** [comment]
Been following this a bit as I am having the same issue reported in #397 using the vanilla  { Link } from "react-router-dom" producing the `<a> cannot appear as a descendant of <a>.` using 0.3.5. 

Some of this conversation is pushing my boundaries of component rendering :) but just wanted to add a note that the issue doesn't appear to be isolated to the NextJs Link component. Not sure if there are any nuances with how NextJs extends from the core React lib or if it even does at all but I've noticed that the NextJs link shows to use `href` in the docs where the react router specifies `to`

```
        <Sidebar.Items className="pt-8 pr-12">
          <Sidebar.ItemGroup>
                  <Sidebar.Item className="mb-2" icon={IoMdPin}>
                      <Link to="/" >
                      Map
                      </Link>
                  </Sidebar.Item>
            </Sidebar.ItemGroup>
        </Sidebar.Items>
```

**Comment by tulup-conner** [comment]
Well, I've brought this up somewhere on this repo before, but the universal standard for component libraries as far as I understand is to provide a prop, generally called `as`, that lets you essentially "cast" any flowbite-react elements that are/can be links to any custom link component. 

```js
<Sidebar.Item className="mb-2" icon={IoMdPin} as={Link} to="/">Map</Sidebar.Item>
```

I tried to do this a few months ago but I'm too stupid. Or, perhaps I just need to try again. Anyway, I think this is probably the most familiar and ergonomic solution.

**Comment by rluders** [comment]
I'm just wondering what would happen if the `Sitebar.Item` became a simple `div` if there is no `href` property attached to it, and if there is a `href` it should become an `a` otherwise, it will use the `children`. So, in this case, could be a `Link` or a `Button` or whatever fits to the desired action.

And all that I'm saying right now could be really dumb... I'm just wondering what could happen. 🤔

**Comment by tulup-conner** [comment]
That makes sense to me, too. I have a theory if we could figure out how to do the `as={}` it might be less code/more readable, but like I said I failed before.  I'll look into this.

**Comment by rluders** [comment]
I'm not sure if I would take my suggestion seriously. I have no idea how it is going to work and if it is a good practice at all. It was just a crazy idea. 🤣

I think that also using the `as={}` could be interesting for `Button` component and so many other components that may need nome "tag mutation".

**Comment by tulup-conner** [comment]
Ok, I tried again and the issue I was having with Sidebar.Item is that you can't override className, which is how Remix facilitates self-aware links via NavLink. That's because we designate it ourselves in Sidebar.Item on L62. It looks like all other properties work just fine.

I'm not sure how to work around this - it might be that I will just need to patch Remix on my apps. I'm not sure if it is something that will come up with other users/use cases where they need to override className for the as={} component and can't.

**Comment by tulup-conner** [comment]
> Ok, I tried again and the issue I was having with Sidebar.Item is that you can't override className, which is how Remix facilitates self-aware links via NavLink. That's because we designate it ourselves in Sidebar.Item on L62. It looks like all other properties work just fine.
> 
> I'm not sure how to work around this - it might be that I will just need to patch Remix on my apps. I'm not sure if it is something that will come up with other users/use cases where they need to override className for the as={} component and can't.

The workaround looks like...

Patch to `flowbite-react`

```js
-          className={classNames(
-            theme.base,
-            isActive && theme.active,
-            !isCollapsed && isInsideCollapse && theme.collapsed.insideCollapse,
-            className,
-          )}
+          className={
+            typeof className === 'string' || typeof className === 'undefined'
+              ? classNames(
+                  theme.base,
+                  isActive && theme.active,
+                  !isCollapsed && isInsideCollapse && theme.collapsed.insideCollapse,
+                  className,
+                )
+              : className
+          }
```

Custom component for ease of use

```js
export const SidebarNavLink: FC<NavLinkProps & SidebarItemProps> = function ({
  children,
  ...props
}) {
  const theme = useTheme().theme.sidebar.item;

  return (
    <FlowbiteSidebar.Item
      as={NavLink}
      className={({ isActive }: { isActive: boolean }) =>
        classNames(theme.base, isActive && theme.active)
      }
      {...props}
    >
      {children}
    </FlowbiteSidebar.Item>
  );
};
```

Actual links

```js
<Sidebar.NavLink to="/brand/components/tooltip">Tooltip</Sidebar.NavLink>
```

The end product in the final snippet here is an ideal DX from my perspective, but I think the pieces in between can be better.

**Comment by rluders** [comment]
@myasteiner Well, I guess that this now changes the PR general goal. So, maybe we could update the title and the description.

**Review by rluders — APPROVED** [review]

**Pull request merged**